### PR TITLE
Revert map optimizations

### PR DESCRIPTION
**Pull request by arhimondr**
This PR reverts optimizations introduced in https://github.com/prestodb/presto/pull/16027. The optimization causes significant regression in certain cases.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by mbasmanova — APPROVED** [review]

**Review by tdcmeehan — APPROVED** [review]

**Review by arunthirupathi — APPROVED** [review]
My apologies for causing this unintended regression, I did a presto verifier for this run though. I will work on the repro, identify the root cause and fix this issue.

**Comment by arunthirupathi** [comment]
**Presto Query CPU Regression**

TL;DR Presto Map hash compuatation was changed from eager to lazy. When map is used as a part of broad cast join, instead of computing once and broad casted, it was repatedly computed by all the consumers. This regressed the DIRCU for the queries.

**Detailed explanation:**

Maps are ubiquitous in Presto (Query engine, ORC Reader/Writer, Exchange between workers and so on). Presto has two representations for Maps, MapBlock and MapBlockBuilder. MapBlockBuilder builds the MapBlock. Presto MapBlocks besides storing the key/value blocks, can optionally store the open hash table for the key to values. Instead of matching keys one by one, it can use this hash table for the key lookup. Presto Map block seekKey methods relies on the open hash table to lookup. 

Presto MapBlock builds the hash table lazily. MapBlockBuilder always computed them. I modified it to lazy in https://github.com/prestodb/presto/pull/16027 . This caused CPU regressions when map blocks are built and broadcasted.

Consider a classic star schema with 2 tables. Dimension table : Products (ProductId, ProductName) , Fact table: Sales (ProductId, Quanity, Price) and the following query

```
SELECT
    ELEMENT_AT(
        SELECT
            MAP_AGG(ProductId, ProductName)
        FROM Products
    ),
    ProductId
) AS ProductName,
Quantity
FROM Sales
```

The typical query plan Presto generates is

1. Read Products dimension table, build a map with key as ProductId and value as ProductName.  (MAP_AGG)
2. Broadcast the map to the readers of the Sales table. The Sales table will be large so reading will be split to multiple workers. 
3. The workers while reading the Sale table, looks up the ProductId from the Map to get the ProductName.  (ELEMENT_AT)

Previously the hash table was computed in Step 1, Serialized in Step 2 and De-Serialized in Step 3. Now the hash table is computed in Step 3. So instead of computing once, it is done repeatedly by multiple workers. The regression is proportional to the number of workers times number of keys in the map, so it increased very quickly.

Some of the regressed queries I looked at, had multiple dimension tables and all of them had few hundred thousand rows to few million rows in the dimension table.

**Why was the hash table computation made lazy:** 

Recently I am working on a change to reuse the ORC reader/writer library from Presto and replace one of our custom implementation for ORC reader/writer. ORC writer does not rely on the hashes and it only relies on the Key/Value blocks. Without the laziness it causes 5% regression on some Map with String columns to be written to the Warehouse. 

**Next steps:**

Before serializing the map block builder, always compute the hash table. There could be other fixes here, but I will discuss with the experts to get the opinion. I will also work with the experts in enabling all the reverted PRs. The other 2 PRs does not have issue from the data I looked at.

P.S: Another unrelated PR to use Segmented Slice in String dictionary writer was also reverted. I looked into all the regressions reported and TableWriteOperator did not regress. Recently I improved the dictionary writer performance by more than 30%. The segmented slice builder in the worst of the cases can regress by 5%, but it saves Out Of Memory errors on the writer when using large dictionary sizes, which is an important use case for us.
